src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
 
         protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object> output)
         {
-            // pass bytes to SslStream through input -> trigger HandleSslStreamRead. After this call sslStreamReadBuffer may or may not have bytes to read
-            this.mediationStream.AcceptBytes(input);
+            int startOffset = input.ReaderIndex;
+            int endOffset = input.WriterIndex;
+            int offset = startOffset;
+            int totalLength = 0;
 
-            if (!this.EnsureAuthenticated())
+            // If we calculated the length of the current SSL record before, use that information.
+            if (this.packetLength > 0)
             {
-                return;
+                if (endOffset - startOffset < this.packetLength)
+                {
+                    return;
+                }
+                else
+                {
+                    offset += this.packetLength;
+                    totalLength = this.packetLength;
+                    this.packetLength = 0;
+                }
+            }
+
+            bool nonSslRecord = false;
+
+            while (totalLength < TlsUtils.MAX_ENCRYPTED_PACKET_LENGTH)
+            {
+                int readableBytes = endOffset - offset;
+                if (readableBytes < TlsUtils.SSL_RECORD_HEADER_LENGTH)
+                {
+                    break;
+                }
+
+                int packetLength = TlsUtils.GetEncryptedPacketLength(input, offset);
+                if (packetLength == -1)
+                {
+                    nonSslRecord = true;
+                    break;
+                }
+
+                Contract.Assert(packetLength > 0);
+
+                if (packetLength > readableBytes)
+                {
+                    // wait until the whole packet can be read
+                    this.packetLength = packetLength;
+                    break;
+                }
+
+                int newTotalLength = totalLength + packetLength;
+                if (newTotalLength > TlsUtils.MAX_ENCRYPTED_PACKET_LENGTH)
```

Comment:
so let's say we read 1 packet and then try to read the 2nd packet, but it exceeds the max_encrypted_packet_length?
don't we need to update this.packetLength? Or do we do that later in this method?